src/util/strencodings.cpp

```diff
@@ -144,40 +144,23 @@ std::optional<std::vector<unsigned char>> DecodeBase64(const char* p)
         -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1
     };
 
-    const char* e = p;
-    std::vector<uint8_t> val;
-    val.reserve(strlen(p));
-    while (*p != 0) {
-        int x = decode64_table[(unsigned char)*p];
-        if (x == -1) break;
-        val.push_back(uint8_t(x));
-        ++p;
-    }
+    if (str.size() % 4 != 0) return {};
+    /* One or two = characters at the end are permitted. */
+    if (str.size() >= 1 && str.back() == '=') str.remove_suffix(1);
+    if (str.size() >= 1 && str.back() == '=') str.remove_suffix(1);
```

Comment:
If true, it could make sense to mention this behaviour change in the commit description (or pull request description).

Reply:
Interesting! The answer is that it's complicated.

So, the test code you suggest here actually passes, but only because it isn't checking the invalid output flag. Current master, I believe, decodes input which isn't a multiple of 4 characters correctly (as if it had padding) but does set the `*pf_invalid` flag.

So this is indeed a behavior change, but only for call sites which don't check the invalid flag. I think there is only one such site: `RPCAuthorized()` in httprpc.cpp, and I think that's a bug (the HTTP spec RFC 7617 for Basic authentication specifies the use of padded Base64 on very casual reading).

If so, the commit that switches to `std::optional`-wrapped results actually fixes this.

Thoughts?

Reply:
Oh right. The commit that changes behaviour is likely the std::optional one (889a5f107c5187c9232f340adfaedfe1b90361d5), not the string_view one I mentioned earlier.

I guess there is nothing needed to be done here. I just wanted to mention the behaviour change here, so that we aren't surprised when someone creates a bug report "My broken javascript rpc library doesn't work".

Reply:
Moved the behavior-changing aspect to a separate commit.